### PR TITLE
Map customer code to a metadata field in the Taxes App

### PR DESCRIPTION
**Pull request by peelar**
## Context

Taxes App synchronizes a group of fields from Saleor to an AvaTax transaction. Some of these fields are used to determine the tax value (e.g. the customer code).

The general approach is to use the Saleor field that matches the purpose of the AvaTax field the most. However, users sometimes want to override that value with their own. This usually comes into play when multiple other systems (where e.g. one system manages the customers) collaborate on the tax calculation. 

## Scope of the PR

- Refactored the app to use `UserFragment` everywhere
- Added new logic to resolving AvaTax customer code. The app now tries to read it from metadata. If not found, it falls back to the user id. When no user id is found, it falls back to the dummy value (as AvaTax docs suggest).

## Checklist

- [x] I added changesets and [read good practices](/.changeset/README.md).
- [x] [I documented it](https://github.com/saleor/saleor-docs/pull/1078)

## Note
This PR isn't breaking. If the metadata field is not found, the app will fall back as usual. We will add the migrations next week.

### CODE REVIEW

**Comment by lkostrowski** [comment]
Please remember about the docs (or add in issue to update them later)

File: apps/taxes/src/modules/avatax/avatax-customer-code-resolver.ts [hunk 1]
```diff
@@ -1,18 +1,32 @@
...
-    return payload.taxBase.sourceObject.user?.id ?? this.fallbackCustomerCode;
-  }
-}
+export const avataxCustomerCode = {
+  resolve(user: UserFragment | null | undefined): string {
+    // see: docs.saleor.io/docs/3.x/developer/app-store/apps/taxes/avatax/overview#customer-code
+    const avataxCustomerCode = user?.avataxCustomerCode;
+
+    if (avataxCustomerCode) {
+      logger.trace(`Returning customer code found in the user metadata.`);
```
Comment (witoszekdev) [inline apps/taxes/src/modules/avatax/avatax-customer-code-resolver.ts:18]: suggestion: Let's add the customer code to logs for debugging purposes

```suggestion
      logger.trace({ avataxCustomerCode }, `Returning customer code found in the user metadata.`);
```

File: apps/taxes/src/modules/avatax/avatax-customer-code-resolver.ts [hunk 2]
```diff
@@ -1,18 +1,32 @@
...
+      return avataxCustomerCode;
+    }
+
+    const userId = user?.id;
+
+    if (userId) {
+      logger.trace(`Returning user id as a customer code.`);
```
Comment (witoszekdev) [inline apps/taxes/src/modules/avatax/avatax-customer-code-resolver.ts:25]:
```suggestion
      logger.trace({ userId }, `Returning user id as a customer code.`);
```

**Review by witoszekdev — APPROVED** [review]

File: apps/taxes/src/modules/avatax/avatax-entity-type-matcher.ts [hunk 3]
```diff
@@ -27,6 +27,7 @@ export class AvataxEntityTypeMatcher {
     return result.value?.[0].code || this.returnFallback();
   }
 
+  // Now that we get the customer code from user metadata, maybe we should get the entity code from the same place?
```
Comment (krzysztofzuraw) [inline apps/taxes/src/modules/avatax/avatax-entity-type-matcher.ts:30]: **Question:** should it be marked as TODO?

**Review by krzysztofzuraw — APPROVED** [review]

**Review by Cloud11PL — APPROVED** [review]

**Pull request merged**